platform-ajax/src/main/typescript/com/softicar/platform/ajax/engine/event/KeyboardEvents.ts

```diff
@@ -0,0 +1,144 @@
+		}
+		return handler;
+	}
+	
+	private getKey(eventName: string): number {
+		switch(eventName) {
+			case 'ENTER':  return DOM_VK_ENTER;
+			case 'ESCAPE': return DOM_VK_ESCAPE;
+			case 'SPACE':  return DOM_VK_SPACE;
+			case 'TAB':    return DOM_VK_TAB;
+		}
+		throw new Error("Internal error: Unsupported keyboard event name.");
+	}
+}
+
+const KEYBOARD_EVENT_MANAGER = new KeyboardEventManager();
+
+class KeyboardEventHandler {
+	private readonly node;
+	private readonly keyCodes = new Map<number, string>();
+	private readonly fireOnKeyUp = new Map<number, boolean>();
+	private readonly preventDefault = new Map<number, boolean>();
+	private readonly cssClassApplier = new Map<number, CssClassApplier>();
+	private lastKeyDown = 0;
+
+	public constructor(node: HTMLElement) {
+		this.node = node;
+	}
+	
+	public install() {
+		if(!this.node.onkeydown && !this.node.onkeyup) {
+			this.node.onkeydown = event => this.handleKeyDown(event);
+			this.node.onkeyup = event => this.handleKeyUp(event);
+		} else {
+			console.log('warning: skipped installation of keyboard event listeners');
+		}
+		return this;
+	}
+
+	public setListenTo(key: number, eventName: string, enabled: boolean) {
+		if(enabled) {
+			this.keyCodes.set(key, eventName);
+			this.preventDefault.set(key, true);
+		} else {
+			this.keyCodes.delete(key);
+		}
+	}
+
+	public setFireOnKeyUp(key: number, enabled: boolean) {
+		this.fireOnKeyUp.set(key, enabled);
+	}
+
+	public setPreventDefault(key: number, enabled: boolean) {
+		this.preventDefault.set(key, enabled);
+	}
+
+	public setCssClassApplier(key: number, applier: CssClassApplier) {
+		this.cssClassApplier.set(key, applier);
+	}
+
+	private handleKeyDown(event: KeyboardEvent) {
+		let eventName = this.keyCodes.get(event.keyCode);
+		if(eventName) {
+			if(!this.fireOnKeyUp.get(event.keyCode) && !event.repeat) {
+				sendOrDelegateEvent(this.node, event, eventName);
+			}
+			let applier = this.cssClassApplier.get(event.keyCode);
+			if(applier) {
+				applier.addClasses();
+			}
+			this.stopFurtherHandling(event);
+		}
+		this.lastKeyDown = event.keyCode;
+	}
+
+	private handleKeyUp(event: KeyboardEvent) {
+		var eventName = this.keyCodes.get(event.keyCode);
+		if(eventName) {
+			if(this.fireOnKeyUp.get(event.keyCode) && this.lastKeyDown == event.keyCode) {
+				sendOrDelegateEvent(this.node, event, eventName);
+			}
+			let applier = this.cssClassApplier.get(event.keyCode);
+			if(applier) {
+				applier.removeClasses();
+			}
+			this.stopFurtherHandling(event);
+		}
+		this.lastKeyDown = 0;
+	}
+
+	private stopFurtherHandling(event: KeyboardEvent) {
+		event.stopPropagation();
+		if(this.preventDefault.get(event.keyCode)) {
+			event.preventDefault();
+		}
+	}
+}
+
+class CssClassApplier {
+	private readonly node: HTMLElement;
+	private readonly classes: string[];
+
+	public constructor(node: HTMLElement, classes: string) {
+		this.node = node;
+		this.classes = classes != ""? classes.split(" ") : [];
```

Comment:
```suggestion
		this.classes = classes != "" ? classes.split(" ") : [];
```
...equal spacing around both "?" and ":".

Reply:
in Java we have no space before the `?` so I did the same here

Reply:
Indeed, I never realized this. Never mind.